gadget/gadget.go

```diff
@@ -716,59 +814,86 @@ func LoadDiskVolumesDeviceTraits(dir string) (map[string]DiskVolumeDeviceTraits,
+		// that a given disk "has" such raw structures at particular
+		// locations, aside from potentially reading and comparing the bytes
+		// at the expected locations, but that is probably fragile and very
+		// non-performant.
+		if !vs.IsPartition() {
+			// skip trying to find non-partitions on disk, it won't work
+			continue
+		}
+
+		devNode, err := MaybeDeviceForStructure(&vs)
+		if err != nil {
+			return "", err
+		} else if devNode != "" {
+			return devNode, nil
+		}
+	}
+	return "", nil
+}
+
 // AllDiskVolumeDeviceTraits takes a mapping of volume name to Volume
 // and produces a map of volume name to DiskVolumeDeviceTraits. Since
 // doing so uses DiskVolumeDeviceTraitsForDevice, it will also
 // validate that disk devices identified for the volume are compatible
 // and matching before returning.
 func AllDiskVolumeDeviceTraits(allVols map[string]*Volume, optsPerVolume map[string]*DiskVolumeValidationOptions) (map[string]DiskVolumeDeviceTraits, error) {
 	// build up the mapping of volumes to disk device traits
-
-	allTraits := map[string]DiskVolumeDeviceTraits{}
-
 	// find all devices which map to volumes to save the current state of the
 	// system
+	allTraits := map[string]DiskVolumeDeviceTraits{}
 	for name, vol := range allVols {
 		// try to find a device for a structure inside the volume, we have a
 		// loop to attempt to use all structures in the volume in case there are
 		// partitions we can't map to a device directly at first using the
 		// device symlinks that FindDeviceForStructure uses
-		dev := ""
-		for _, vs := range vol.Structure {
-			// TODO: This code works for volumes that have at least one
-			// partition (i.e. not type: bare structure), but does not work for
-			// volumes which contain only type: bare structures with no other
-			// structures on them. It is entirely unclear how to identify such
-			// a volume, since there is no information on the disk about where
-			// such raw structures begin and end and thus no way to validate
-			// that a given disk "has" such raw structures at particular
-			// locations, aside from potentially reading and comparing the bytes
-			// at the expected locations, but that is probably fragile and very
-			// non-performant.
-
-			if !vs.IsPartition() {
-				// skip trying to find non-partitions on disk, it won't work
-				continue
-			}
-
-			structureDevice, err := FindDeviceForStructure(&vs)
-			if err != nil && err != ErrDeviceNotFound {
-				return nil, err
-			}
-			if structureDevice != "" {
-				// we found a device for this structure, get the parent disk
-				// and save that as the device for this volume
-				disk, err := disks.DiskFromPartitionDeviceNode(structureDevice)
-				if err != nil {
-					return nil, err
-				}
-
-				dev = disk.KernelDeviceNode()
-				break
-			}
-		}
-
-		if dev == "" {
+		dev, err := MaybeDeviceForVolume(vol)
+		if err != nil {
+			return nil, err
+		} else if dev == "" {
```

Comment:
nitpick: the "else" is a bit confusing as you will return if there is an error, I'd prefer `if dev == ...` directly

Reply:
okay this is a good point - fixed